doc/sphinx/practical-tools/utilities.rst

```diff
@@ -190,6 +190,85 @@ since it results in a “packed” plugin: All auxiliary modules (as
+outside a quoted string, and preceded by whitespace or a closing double quote.
+This means ``Filename#.v`` and ``"Filename With # Spaces.v"`` won’t cause
+comments. The ``#`` is in these cases treated like an ordinary character.
+
+.. _grammar_of_coqproject:
+
+The grammar of _CoqProject
+++++++++++++++++++++++++++
+The parser of coq_makefile uses the following grammar:
+
+  .. insertprodn CoqProject unquoted_string
+
+  .. prodn::
+     CoqProject ::= {* {| @blank | @comment | @quoted_string | @unquoted_string } }
+     blank ::= {| space | horizontal_tab | newline }
+     comment ::= # {* comment_byte } newline
+     quoted_string ::= " {* quoted_string_byte } "
+     unquoted_string ::= string_start_byte {* unquoted_string_byte }
+
+Where the following definitions apply:
+
+* :n:`space`, :n:`horizontal_tab` and :n:`newline` stand for the corresponding ASCII values.
+* :n:`comment_byte` is the set of all bytes except :n:`newline`.
+* :n:`quoted_string_byte` is the set of all bytes except ``"``.
+* :n:`string_start_byte` is the set of all bytes except those that match :n:`@blank`, or are ``"`` or ``#``.
+* :n:`unquoted_string_byte` is the set of all bytes except those that match :n:`@blank`.
+
+The parser produces a list of strings in the same order as they were
+encountered in ``_CoqProject``. All :n:`@blank` and :n:`@comment` tokens
+get ignored and the double quotes of :n:`@quoted_string` tokens are removed as
+well.
+
+The list of strings is then parsed using the following grammar:
+
+  .. insertprodn options option3
+
+  .. prodn::
+     options ::= {* @option }
+     option ::= {| @option0 | @option1 string | @option2 string string | @option3 string string string | string = string | -arg string | string }
+     option0 ::= {| -impredicative-set | -no-install | -custom | -no-opt | -byte | -full | -opt }
+     option1 ::= {| -install | -native-compiler | -f | -o }
+     option2 ::= {| -Q | -I | -R }
+     option3 ::= {| -extra | -extra-phony }
+
+The grammar given above might accept a given list of strings in two different
+ways, precisely if an equals-sign is involved. In this case the implementation
+chooses the leftmost interpretation.
+:n:`string` stands for an arbitrary string and all other strings stand for
+themselves.
+
+The semantics of ``-arg`` are as follows: all strings that were found
+immediately to the right of ``-arg`` by the above grammars are concatenated
+with single spaces between them. The resulting string is written in
+``CoqMakefile.conf`` as variable ``COQMF_OTHERFLAGS``. This variable is then
+passed to ``coqc`` using something similar to the following invocation: ``coqc
+$(COQMF_OTHERFLAGS)``.
+
+From this intermediate step in the ``CoqMakefile.conf``, a lot of the
+peculiarities of ``-arg`` arise, since the string may be split appart
+differently than how coq_makefile did it, variables like ``$(MAKE)`` can get
+expanded and single quotes mark where no expansion or splitting should take
+place.
```

Comment:
This is not too clear and not written from the user's perspective.

Reply:
I think this PR should also change coq_makefile so that stuff like `-arg "some stuff"` is immediately desugared to `-arg some -arg stuff` so that this paragraph about nasty peculiarities can be omitted. The quotation rules will then be specific to coq_makefile, and will put make and dune out of the picture.

Reply:
> This is not too clear and not written from the user's perspective.

I agree and I’ll think about this, how it could be made easier to understand. Trouble is, that the current behaviour *just is* difficult to understand.

Reply:
@gares, how would you expect `-arg "some 'weird stuff' happens"` to behave? I think we clearly want to parse this as `["-arg", "some 'weird stuff' happens"]` (in coq_makefile, before splitting), but a naive implementation would pass the following list to coqc (after splitting): `["some", "'weird" "stuff'", "happens"]`, but I think `["some", "weird stuff", "happens"]` is more useful.

We could define, that single quotes only have effect inside the argument to `-arg`.

Reply:
It's a judgement call whether and how to document bugs and limitations.  It's better to fix simple bugs rather than documenting them, but that's not always easy.  On the other hand, we do want to document limitations that may not be addressed soon (or ever).  Also, we don't want explanations of bugs and limitations to overshadow the basic description of the feature.

It would be great if you're willing to clean up the argument parsing a bit so it makes sense and is easier to describe and use.  For example, addressing the point you make [below ](https://github.com/coq/coq/pull/14558#issuecomment-869660724).

Reply:
So am I free to introduce, say, backslash escaping and propose a new parsing/output algorithm?

Reply:
Feel free to propose any changes you wish, but they will have to be approved by reviewer(s).  Backslash escaping sounds like a small and compatible improvement--maybe just go ahead and code that.  Proposing a new parsing/output algorithm you might describe it first before writing code.